### PR TITLE
coverage: Use `SpanMarker` to improve coverage spans for `if !` expressions

### LINKED ISSUE #115468: Coverage instrumentation sometimes thinks the negation operator isn't executable
#111752 introduced this behaviour in coverage reports:

```
   LL|       |    if
   LL|       |        !
   LL|      1|        is_true
   LL|      0|    {
```

Notice that the line with `!` no longer has a coverage count, even though it should have the same count as `is_true`.

---

This is caused by the fact that MIR building specifically handles `if !` by flipping the then/else arms, so the original condition expression (containing `!`) does not have its span represented by any statement in MIR.

### PR DESCRIPTION
**Pull request by Zalathar**
Coverage instrumentation works by extracting source code spans from MIR. However, some kinds of syntax are effectively erased during MIR building, so their spans don't necessarily exist anywhere in MIR, making them invisible to the coverage instrumentor (unless we resort to various heuristics and hacks to recover them).

This PR introduces `CoverageKind::SpanMarker`, which is a new variant of `StatementKind::Coverage`. Its sole purpose is to represent spans that would otherwise not appear in MIR, so that the coverage instrumentor can extract them.

When coverage is enabled, the MIR builder can insert these dummy statements as needed, to improve the accuracy of spans used by coverage mappings.

Fixes #115468.

---

@rustbot label +A-code-coverage

### CODE REVIEW

**Comment by Zalathar** [comment]
It occurs to me that another approach would have been to make the `UnOp::Not` match arm conditional on `!this.tcx.sess.instrument_coverage()`.

That would avoid the need to inject a span marker in this case, though I still think `continue` justifies having span markers on its own.

(And the downside would be that coverage builds wouldn't get to omit the temporary boolean value.)

**Comment by Zalathar** [comment]
> It occurs to me that another approach would have been to make the `UnOp::Not` match arm conditional on `!this.tcx.sess.instrument_coverage()`.
> 
> That would avoid the need to inject a span marker in this case, though I still think `continue` justifies having span markers on its own.
> 
> (And the downside would be that coverage builds wouldn't get to omit the temporary boolean value.)

Looking again at the thread for #111752, I realise that some of those match arms are now load-bearing for language semantics. I think that mostly applies to the `&&`/`||` arms, but overall it means I'd have to be pretty careful about that sort of thing.

**Review by Swatinem — APPROVED** [review]
you mentioned that you plan to eventually move capturing this span information to HIR, so I think injecting these coverage statements here is just temporary?

**Comment by Zalathar** [comment]
I actually think of these span markers as mostly separate from the ones that I intend to introduce for capturing branch spans.

This PR was the result of a few different ideas coming together in my head:
- I had been experimenting with marker statements for branch coverage, so the idea of injecting marker statements was on my mind.
- I remembered that there is pre-existing code in MIR building that injects dummy assignment statements for `continue` expressions, so that they can be picked up properly by the coverage instrumentor.
- I realised that introducing `SpanMarker` would let me clean up the `continue` handling a bit (by not having to inject “real” dummy statements), while also letting me solve #115468.

So I think these particular marker statements will hang around even after we also have proper branch detection in MIR building. These span markers are mainly about giving more accurate hints to the existing span-extraction code in the instrumentor, whereas the branch markers will be separate from that span-extraction stuff.

**Comment by Zalathar** [comment]
One distinction to note is that these span markers get injected into the basic block that represents the *condition*, whereas branch markers will get injected into the blocks representing the then/else arms.

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 1]
```diff
@@ -90,6 +90,11 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 let local_scope = this.local_scope();
                 let (success_block, failure_block) =
                     this.in_if_then_scope(local_scope, expr_span, |this| {
+                        // Help out coverage instrumentation by injecting a dummy statement with
+                        // the original condition's span (including `!`). This fixes #115468.
+                        if this.tcx.sess.instrument_coverage() {
```
Comment (cjgillot) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:95]: Should this test happen inside `push_coverage_span_marker`?

File: compiler/rustc_mir_build/src/build/scope.rs [hunk 2]
```diff
@@ -723,14 +723,6 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
     }
 
-    // Add a dummy `Assign` statement to the CFG, with the span for the source code's `continue`
-    // statement.
-    fn add_dummy_assignment(&mut self, span: Span, block: BasicBlock, source_info: SourceInfo) {
-        let local_decl = LocalDecl::new(Ty::new_unit(self.tcx), span);
-        let temp_place = Place::from(self.local_decls.push(local_decl));
-        self.cfg.push_assign_unit(block, source_info, temp_place, self.tcx);
-    }
-
```
Comment (cjgillot) [inline compiler/rustc_mir_build/src/build/scope.rs:733]: Nice :tada:

**Review by cjgillot — COMMENTED** [review]
r=me with 2 nits.

File: compiler/rustc_mir_build/src/build/matches/mod.rs (hunk 1 above)
Comment (Zalathar) [reply compiler/rustc_mir_build/src/build/matches/mod.rs:95]: `push_coverage_span_marker` is a method on `CFG`, which doesn't have its own copy of `tcx` or `sess`.

So the only way to perform the test would be to also pass tcx/sess as an additional argument, which seems backwards to me. Better to perform the test in the outer code that already has access to the necessary information.

Having the test outside also makes it easier to see from outside that this code has no effect on the CFG in non-coverage builds.

**Comment by Zalathar** [comment]
This turns out to have contributed to the emergence of #118850.

**Pull request merged**